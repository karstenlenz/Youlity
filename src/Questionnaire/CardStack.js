import React from 'react'
import styled from 'styled-components/macro'
import SwipeableCard from './SwipeableCard'
import PropTypes from 'prop-types'

CardStack.propTypes = {
  questions: PropTypes.array.isRequired,
  currentQuestionIndex: PropTypes.number.isRequired,
  handleAnswer: PropTypes.func.isRequired,
  offset: PropTypes.number.isRequired,
  setOffset: PropTypes.func.isRequired,
  swipeThreshold: PropTypes.number.isRequired,
}

export default function CardStack({
  questions,
  currentQuestionIndex,
  handleAnswer,
  offset,
  setOffset,
  swipeThreshold,
}) {
  return (
    <CardStackContainer>
      <BackgroundCardsContainer>
        {questions[currentQuestionIndex + 2] && (
          <Card scale={0.9} top="-25px" brightness="96%" zIndex="1">
            {questions[currentQuestionIndex + 2]}
          </Card>
        )}
        {questions[currentQuestionIndex + 1] && (
          <Card scale={0.95} top="-15px" brightness="98%" zIndex="2">
            {questions[currentQuestionIndex + 1]}
          </Card>
        )}
      </BackgroundCardsContainer>
      <SwipeableCard
        handleAnswer={handleAnswer}
        offset={offset}
        setOffset={setOffset}
        swipeThreshold={swipeThreshold}
      >
        {questions[currentQuestionIndex]}
      </SwipeableCard>
    </CardStackContainer>
  )
}

const CardStackContainer = styled.div`
  position: relative;
  height: 200px;
`
const BackgroundCardsContainer = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
`

const Card = styled.div`
  background: white;
  border-radius: 12px;
  box-shadow: var(--primary-shadow);
  width: 100%;
  max-width: 72.5%;
  min-height: 200px;
  margin: auto;
  padding: 20px;
  font-size: 1em;
  position: absolute;
  left: 50%;
  transform: ${(props) => 'translateX(-50%) scale(' + props.scale + ');'};
  top: ${(props) => props.top};
  z-index: ${(props) => props.zIndex};
  filter: brightness(${(props) => props.brightness});
`
