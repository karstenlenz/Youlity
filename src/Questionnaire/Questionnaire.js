import React, { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { personalityStyleData } from '../data/personalityStyleData'

export default function Questionnaire({ testIds }) {
  const history = useHistory()

  const [questionRound, setQuestionRound] = useState(0)
  const currentTestIndex = testIds[questionRound] - 1
  const questions = personalityStyleData[currentTestIndex]?.questions ?? [
    'Brauchen Sie in hohem Maße Lob und Anerkennung?',
    'Möchten Sie besser sein als andere?',
    'Reagieren Sie empfindlich auf Kritik, selbst wenn diese berechtigt ist?',
    'Haben Sie ab und zu Phasen, in denen Sie an Ihren Fähigkeiten, Erfolgen etc. zweifeln?',
    'Erleben Sie Phasen, in denen Sie sehr zufrieden mit sich sind und denken, dass Sie gut sind?',
    'Haben Sie die Tendenz, in besonderer Weise behandelt werden zu wollen?',
    'Haben Sie deutliche Erwartungen, an die sich andere halten sollten, z. B. Sie nicht zu behindern u. a.?',
    'Neigen Sie dazu, andere Personen „einzuspannen“, ihnen Aufgaben zu geben, die Sie eigentlich selbst erledigen sollten?',
  ]
  const [currentQuestionIndex, setCurrentQuestionIndex] = useState(0)
  const [answers, setAnswers] = useState([])
  const [resultUrl, setResultUrl] = useState('/result/')

  useEffect(() => {
    questionRound === 2 && history.push(resultUrl)
  }, [questionRound, resultUrl, history])

  return (
    <>
      <h1>Fragebogen {questionRound + 1} / 2</h1>
      <p>"{personalityStyleData[currentTestIndex]?.name}"</p>
      <h2>
        Frage {currentQuestionIndex + 1} / {questions.length}
      </h2>
      <h3>{questions[currentQuestionIndex]}</h3>
      <button onClick={() => handleAnswer(false)}>(Eher) Nein</button>
      <button onClick={() => handleAnswer(true)}>(Eher) Ja</button>
    </>
  )

  function handleAnswer(answer) {
    if (currentQuestionIndex === questions.length - 1) {
      setQuestionRound(questionRound + 1)
      setResultUrl(
        resultUrl + (currentTestIndex + 1) + countYes([...answers, answer])
      )
      setCurrentQuestionIndex(0)
      setAnswers([])
    } else {
      setAnswers([...answers, answer])
      setCurrentQuestionIndex(currentQuestionIndex + 1)
    }
  }
}

export function countYes(answers) {
  return answers.reduce((yesCount, answer) => {
    return answer ? yesCount + 1 : yesCount
  }, 0)
}
