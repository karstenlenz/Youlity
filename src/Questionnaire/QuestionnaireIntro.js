import React from 'react'
import Button from '../common/Button'
import HeadlineUnderline from '../common/HeadlineUnderline'

export default function QuestionnaireIntro({ onButtonClick }) {
  return (
    <>
      <img alt="" src="/img/questionnaire-intro.svg" />
      <HeadlineUnderline>
        <h1>Schritt 2: Fragebogen</h1>
      </HeadlineUnderline>
      <p>
        Durch die Auswahl Ihrer Bedürfnisse haben wir 2 potentielle
        Persönlichkeitsstile identifiziert. <br /> Bitte füllen Sie jeweils
        einen Fragebogen aus, um die Ausprägung des jeweiligen Stils bei Ihnen
        zu testen. Es werden jeweils 8 Ja/Nein-Fragen gestellt. Pro Bogen dauert
        es ca. 3 Minuten. Die Reaktionszeit spielt keine Rolle, Sie können den
        Bogen in Ruhe ausfüllen.
      </p>
      <Button onClick={onButtonClick}>Weiter zum Fragebogen</Button>
    </>
  )
}
