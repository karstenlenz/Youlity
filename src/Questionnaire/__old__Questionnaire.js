import React, { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { personalityStyleData } from '../data/personalityStyleData'
import HeadlineUnderline from '../common/HeadlineUnderline'
import Button from '../common/Button'
import styled from 'styled-components/macro'

export default function Questionnaire({ testIds }) {
  const history = useHistory()

  const [questionRound, setQuestionRound] = useState(0)
  const currentTestIndex = testIds[questionRound] - 1
  const questions = personalityStyleData[currentTestIndex]?.questions ?? []
  const [currentQuestionIndex, setCurrentQuestionIndex] = useState(0)
  const [answers, setAnswers] = useState([])
  const [resultUrl, setResultUrl] = useState('/result/')

  useEffect(() => {
    questionRound === 2 && history.push(resultUrl)
  }, [questionRound, resultUrl, history])

  return (
    <>
      <img alt="" src="/img/questionnaire-intro.svg" />
      <HeadlineUnderline>
        <h1>Fragebogen {questionRound + 1} / 2</h1>
      </HeadlineUnderline>
      <h2>
        Frage {currentQuestionIndex + 1} / {questions.length}
      </h2>
      <h3>{questions[currentQuestionIndex]}</h3>

      <ButtonRow>
        <Button
          btnType="white"
          width="47.5"
          onClick={() => handleAnswer(false)}
        >
          (Eher) Nein
        </Button>
        <Button btnType="white" width="47.5" onClick={() => handleAnswer(true)}>
          (Eher) Ja
        </Button>
      </ButtonRow>
    </>
  )

  function handleAnswer(answer) {
    if (currentQuestionIndex === questions.length - 1) {
      setQuestionRound(questionRound + 1)
      setResultUrl(
        resultUrl + (currentTestIndex + 1) + countYes([...answers, answer])
      )
      setCurrentQuestionIndex(0)
      setAnswers([])
    } else {
      setAnswers([...answers, answer])
      setCurrentQuestionIndex(currentQuestionIndex + 1)
    }
  }
}

export function countYes(answers) {
  return answers.reduce((yesCount, answer) => {
    return answer ? yesCount + 1 : yesCount
  }, 0)
}

const ButtonRow = styled.section`
  display: flex;
  justify-content: space-between;
`
