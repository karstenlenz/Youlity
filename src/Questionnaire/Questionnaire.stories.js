import React from 'react'
import Questionnaire from './Questionnaire'

export default {
  title: 'Questionnaire',
  component: Questionnaire,
}

export const Narzissmus = () => <Questionnaire />
