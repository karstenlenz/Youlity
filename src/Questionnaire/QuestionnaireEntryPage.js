import React from 'react'
import { useParams } from 'react-router-dom'
import Questionnaire from './Questionnaire'

export default function QuestionnaireEntryPage() {
  const { testIdParam } = useParams()
  const testIds = testIdParam.split('').map((id) => parseInt(id))

  return <Questionnaire testIds={testIds} />
}
