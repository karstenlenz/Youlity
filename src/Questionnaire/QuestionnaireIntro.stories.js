import React from 'react'
import QuestionnaireIntro from './QuestionnaireIntro'
import { BrowserRouter as Router } from 'react-router-dom'

export default {
  title: 'Questionnaire',
  component: QuestionnaireIntro,
}

export const QuestionnaireIntroComponent = () => (
  <Router>
    <QuestionnaireIntro />
  </Router>
)
