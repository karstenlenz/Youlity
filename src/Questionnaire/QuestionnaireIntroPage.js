import React from 'react'
import { useParams } from 'react-router-dom'
import QuestionnaireIntro from './QuestionnaireIntro'

export default function QuestionnaireIntroPage() {
  const { questionnaireIds, results } = useParams()

  return (
    <QuestionnaireIntro questionnaireIds={questionnaireIds} result={results} />
  )
}
