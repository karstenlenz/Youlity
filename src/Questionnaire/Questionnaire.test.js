import { render } from '@testing-library/react'
import React from 'react'
import Questionnaire, { countYes } from './Questionnaire'

describe('The questionnaire component', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Questionnaire
        userPersonalityStyleIds={[3, 7, 5]}
        onQuestionnaireEnd={jest.fn}
        round={0}
      />
    )
    expect(container).toMatchSnapshot()
  })
})

describe('countYes', () => {
  it('counts all the yesses', () => {
    const testAnswers = [true, true, true, true, true, true, true, true]
    const result = countYes(testAnswers)
    expect(result).toEqual(8)
  })

  it('doesnt count all the noes', () => {
    const testAnswers = [false, false, false, false, false, false, false, false]
    const result = countYes(testAnswers)
    expect(result).toEqual(0)
  })

  it('only counts the yesses', () => {
    const testAnswers = [false, true, false, true, true, false, false, false]
    const result = countYes(testAnswers)
    expect(result).toEqual(3)
  })
})
