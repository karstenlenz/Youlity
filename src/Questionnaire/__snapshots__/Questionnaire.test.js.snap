// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The questionnaire component renders correctly 1`] = `
<div>
  <img
    alt=""
    src="/img/questionnaire-intro.svg"
  />
  <section
    class="HeadlineUnderline__HeadlineContainer-sc-1reeb6v-0 jWFWIm"
  >
    <h1>
      Fragebogen 
      1
       / 2
    </h1>
    <svg
      class="HeadlineUnderline__UnderlineStyled-sc-1reeb6v-1 hZOZZy"
    >
      underline.svg
    </svg>
  </section>
  <h2
    class="Questionnaire__H2Styled-sc-13hxlgu-1 kfJYVs"
  >
    Frage 
    1
     / 
    8
  </h2>
  <div
    class="CardStack__CardStackContainer-vgb89h-0 byiENK"
  >
    <div
      class="CardStack__BackgroundCardsContainer-vgb89h-1 MfyjY"
    >
      <div
        class="CardStack__Card-vgb89h-2 kyUaud"
        scale="0.9"
      >
        Reagieren Sie empfindlich auf Kritik, selbst wenn diese berechtigt ist?
      </div>
      <div
        class="CardStack__Card-vgb89h-2 cydTFS"
        scale="0.95"
      >
        Möchten Sie besser sein als andere?
      </div>
    </div>
    <div>
      <div
        class="SwipeableCard__Card-zqgmcw-0 fZHvwI"
        offset="0"
      >
        Brauchen Sie in hohem Maße Lob und Anerkennung?
      </div>
    </div>
  </div>
  <div
    class="Questionnaire__ArrowContainer-sc-13hxlgu-3 hgOFzU"
  >
    <svg>
      arrow_left.svg
    </svg>
    <svg>
      arrow_right.svg
    </svg>
  </div>
  <section
    class="Questionnaire__ButtonRow-sc-13hxlgu-0 jArQLV"
  >
    <button
      class="Button__BtnStyled-sc-129yfz5-0 ibTELz Questionnaire__QuestionnaireButton-sc-13hxlgu-2 kEMYte"
      width="47.5"
    >
      (Eher) Nein
    </button>
    <button
      class="Button__BtnStyled-sc-129yfz5-0 ibTELz Questionnaire__QuestionnaireButton-sc-13hxlgu-2 kEMYte"
      width="47.5"
    >
      (Eher) Ja
    </button>
  </section>
</div>
`;
